File: backend/internal/subtitle/language_contract_test.go
package subtitle

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"
)

func languageFixturePath(t *testing.T, name string) string {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	// backend/internal/subtitle -> repo root
	root := filepath.Clean(filepath.Join(filepath.Dir(file), "../../.."))
	return filepath.Join(root, "testdata", "language", name)
}

func loadLanguageFixture[T any](t *testing.T, name string) []T {
	t.Helper()
	data, err := os.ReadFile(languageFixturePath(t, name))
	if err != nil {
		t.Fatalf("read fixture %s: %v", name, err)
	}
	var out []T
	if err := json.Unmarshal(data, &out); err != nil {
		t.Fatalf("parse fixture %s: %v", name, err)
	}
	return out
}

func TestContractBilingualFlags(t *testing.T) {
	type row struct {
		Input string `json:"input"`
		Want  bool   `json:"want"`
	}
	for _, tt := range loadLanguageFixture[row](t, "bilingual_flags.json") {
		if got := IsBilingualLanguage(tt.Input); got != tt.Want {
			t.Fatalf("IsBilingualLanguage(%q)=%v want %v", tt.Input, got, tt.Want)
		}
	}
}

func TestContractDetectType(t *testing.T) {
	type row struct {
		Input string `json:"input"`
		Want  string `json:"want"`
	}
	for _, tt := range loadLanguageFixture[row](t, "detect_type.json") {
		if got := DetectLanguageType(tt.Input); got != tt.Want {
			t.Fatalf("DetectLanguageType(%q)=%q want %q", tt.Input, got, tt.Want)
		}
	}
}

func TestContractNameLabels(t *testing.T) {
	type row struct {
		Input     string `json:"input"`
		WantLabel string `json:"wantLabel"`
	}
	for _, tt := range loadLanguageFixture[row](t, "name_labels.json") {
		if got := LabelFromFileName(tt.Input); got != tt.WantLabel {
			t.Fatalf("LabelFromFileName(%q)=%q want %q", tt.Input, got, tt.WantLabel)
		}
	}
}

File: frontend/src/lib/language-contract.test.ts
import { describe, expect, test } from "bun:test";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { detectSubtitleLanguageType, inferUploadLanguageLabel, isBilingualLanguage } from "@/lib/subtitle-language";

const root = join(dirname(fileURLToPath(import.meta.url)), "../../..");

function loadFixture<T>(name: string): T[] {
  const path = join(root, "testdata", "language", name);
  return JSON.parse(readFileSync(path, "utf8")) as T[];
}

describe("language contract fixtures (shared with Go)", () => {
  test("bilingual_flags.json", () => {
    const rows = loadFixture<{ input: string; want: boolean }>("bilingual_flags.json");
    for (const row of rows) {
      expect(isBilingualLanguage(row.input), row.input).toBe(row.want);
    }
  });

  test("detect_type.json", () => {
    const rows = loadFixture<{ input: string; want: string }>("detect_type.json");
    for (const row of rows) {
      expect(detectSubtitleLanguageType(row.input) as string, row.input).toBe(row.want);
    }
  });

  test("name_labels.json", () => {
    const rows = loadFixture<{ input: string; wantLabel: string }>("name_labels.json");
    for (const row of rows) {
      expect(inferUploadLanguageLabel(row.input), row.input).toBe(row.wantLabel);
    }
  });
});
